### PR TITLE
feat(github): Implement ApiCache data structure

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

Create cache datastructure that can be used for any paginated data that have `number` and `updated_at` field.
Doesn't change anything, so it's safe to merge.

## Context

- Ref: #14617

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
In this form it's unused and essentially untested though?

File: lib/modules/platform/github/list-cache.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+export function getEmptyCache<T extends RestPageItem>(): RestPageCache<T> {
+  return {
+    items: {},
+    timestamp: DateTime.fromISO('1900-01-01').toISO(),
+    etag: '',
+  };
+}
+
+export function getItem<T extends RestPageItem>(
```
Comment (viceice) [inline lib/modules/platform/github/list-cache.ts:12]: i think this is a good candidate for a class

Comment (zharinov) [reply lib/modules/platform/github/list-cache.ts]: My point is to attach the cache object to repo cache structure during init phase to mutate it during run:

https://github.com/zharinov/renovate/blob/8391721640901be24b78d67af36ed8122d3abd71/lib/util/cache/repository/types.ts#L56

The finalize phase is unavailable from the platform code, so it seems like we don't have enough control to class serialization/deserialization.

Comment (viceice) [reply lib/modules/platform/github/list-cache.ts]: we could wrap it when getting from cache, so object ref is still hold?

Comment (zharinov) [reply lib/modules/platform/github/list-cache.ts]: Mm, now I get it, good idea 👍

**Comment by zharinov** [comment]
> In this form it's unused and essentially untested though?

Yes, it isn't tested for production yet

File: lib/modules/platform/github/api-cache.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    return this.cache.etag ?? null;
+  }
+
+  setEtag(value: string | null): void {
+    if (value === null) {
+      delete this.cache.etag;
+    } else {
+      this.cache.etag = value;
+    }
+  }
```
Comment (viceice) [inline lib/modules/platform/github/api-cache.ts:17]:
```suggestion
  get etag(): string | null {
    return this.cache.etag ?? null;
  }

  set etag(value: string | null) {
    if (value === null) {
      delete this.cache.etag;
    } else {
      this.cache.etag = value;
    }
  }
```

I mean real getter and setter 😉

File: lib/modules/platform/github/api-cache.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+  /**
+   * @returns Date formatted to use in HTTP headers
+   */
+  getLastModified(): string | null {
+    const { lastModified } = this.cache;
+    return lastModified ? DateTime.fromISO(lastModified).toHTTP() : null;
+  }
```
Comment (viceice) [inline lib/modules/platform/github/api-cache.ts:25]:
```suggestion
  get lastModified(): string | null {
    const { lastModified } = this.cache;
    return lastModified ? DateTime.fromISO(lastModified).toHTTP() : null;
  }
```

File: lib/modules/platform/github/api-cache.ts (hunk 2 above)
Comment (zharinov) [reply lib/modules/platform/github/api-cache.ts:17]: Haha

**Review by viceice — APPROVED** [review]
🤷‍♂️ Not very useful yet 🙃

**Pull request merged**